keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -244,7 +244,226 @@ There are likely others.
+  for now this KEP simply ignores the gap.
+
+One of the biggest questions is how to formulate the scheduling
+parameters.  There are several related concepts in the state of the
+art of scheduling, and we are trying to figure out what to adopt
+and/or invent.  We would prefer to invent as little as possible; it is
+a non-trivial thing to invent a new --- and sufficiently efficient ---
+scheduling algorithm and prove it correct.  The vmware product line
+uses scheduling parameters named "reservation", "limit", and "shares"
+for each class of workload.  The first two are known elsewhere as
+"assured" and "ceiling".  Finding a published algorithm that
+implements all three has not proven easy.  Alternatively, priorities
+are easy to implement and arguably more desirable --- provided there
+is some form of fairness within each priority level.  The current
+thinking is in that direction: use priorities, with simple equal
+fairness among some categories of requests in each priority level.
+There are published scheduling algorithms that provide fairness, and
+we hope to use/adapt one of them to apply independently within the
+confines of each priroity level.
+
+Another issue is whether to manage QPS or concurrency or what.
+Managing QPS leaps first to mind, perhaps because it is a simple
+concept and perhaps because it is familiar from the self-restraint
+that clients apply today.  But we want to also take service time into
+account; a request flow with longer service times should get less QPS
+because its requests are "heavier" --- they impose more load on the
+apiserver.  A natural way to do this is with an inverse linear
+relation.  For example, when two CPU-bound request flows are getting
+equal CPU from the apiserver, and the first flow's requests have a
+service time that is X times the service time of the second flow's
+requests, the first flow's QPS is 1/X of the second's.  This is
+exactly analogous to what happens in networking: if two flows are
+getting the same bandwidth, and one flow's packets are X times as big
+as the second's, then the first flow's packets per second rate is 1/X
+that of the second flow.  This inverse linear relation amounts to
+managing the product of QPS * service time.  That is equivalent to
+managing concurrency.  Managing concurrency is an obvious choice for
+memory, and we now see it is a good choice for CPU too.  This is also
+a convenient choice because it is what the max-in-flight handler is
+doing today, so we would be making a relatively modest extension to
+that hendler's conception.
+
+Compared to traditional scheduling problems, ours is harder because of
+the combination of these facts: (1) (unlike a router handling a
+packet) the apiserver does not know beforehand how long a request will
+take to serve nor how much memory it will consume, and (2) (unlike a
+CPU scheduler) the apiserver can not suspend and resume requests.
+Also, we are really loathe to abort a request once it has started
```

Comment:
> The global timeout we currently have sends a cancellation to the user but no one has yet made it actually cancel the goroutines / etcd requests, so it doesn't actually free up resources.

Agree. But at least some of those is a bug, rather than a feature :-)

> Watch requests are relatively quite safe to cancel early.

I'm not entirely sure I agree - if we cancel a watch, this will probably almost immediately be resend...
I guess my point here is that we need to be very careful what to do if watch, as:
- canceling a watch may not give us a viable gain (it's hard to estimate cost of watch)
- if cancelled and not accepted for quite some time, it will result most probably in relist, so may be even worse...

I'm not saying that we should never cancel watches, but we should do that pretty carefully.

Reply:
Remember that the current goals say we are not (yet) going to attempt to manage WATCH requests.

Reply:
Note also that the goals say that ineffective cancellations are a bug, not a feature for this KEP to deal with.